src/crowdsale/CrowdSaleWithFees.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { CrowdSale, Sale, SaleInfo, AlreadyClaimed, SaleState, BadSaleState } from "./CrowdSale.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+/**
+ * @title CrowdSaleWithFees
+ * @author molecule.to
+ * @notice a plain crowdsale that takes 0.5% fees of the funding goal upon sale settlement
+ */
+contract CrowdSaleWithFees is CrowdSale, Ownable {
+    using SafeERC20 for IERC20Metadata;
+
+    mapping(uint256 => uint256) crowdSaleFees;
+    uint256 public feesPercentage;
+
+    event Started(uint256 indexed saleId, address indexed issuer, Sale sale, uint256 feesPercentage);
+    /**
+     * is called when we deploy this smart contract,
+     * we need to instantiate the initialFees percentage and the contract owner to send the fees to at the end of each successful auction
+     * @param _feesPercentage the percentage of fees to cut of each auction
+     */
+
+    constructor(uint256 _feesPercentage) {
+        feesPercentage = _feesPercentage;
+    }
+
+    function getFees() public view returns (uint256) {
+        return feesPercentage;
+    }
+
+    function getCrowdSaleFees(uint256 saleId) public view returns (uint256) {
+        return crowdSaleFees[saleId];
+    }
+
+    function updateCrowdSaleFees(uint256 newFee) public onlyOwner {
+        feesPercentage = newFee;
+    }
+
+    /**
+     * @notice will instantiate a new crowdsale with fees when none exists yet
+     *
+     * @param sale sale configuration
+     * @return saleId the newly created sale's id
+     */
+    function startSale(Sale calldata sale) public override returns (uint256 saleId) {
+        saleId = super.startSale(sale);
+        crowdSaleFees[saleId] = feesPercentage;
+    }
+
+    function _afterSaleStarted(uint256 saleId) internal override {
+        emit Started(saleId, msg.sender, _sales[saleId], crowdSaleFees[saleId]);
+    }
+
+    function settle(uint256 saleId) public override {
+        Sale storage sale = _sales[saleId];
+        super.settle(saleId);
+        uint256 saleFees = sale.fundingGoal * feesPercentage / 1000;
+        sale.fundingGoal = sale.fundingGoal - saleFees;
+        sale.biddingToken.safeTransfer(owner(), saleFees);
```

Comment:
https://github.com/moleculeprotocol/IPNFT/blob/main/src/crowdsale/CrowdSale.sol#L203C13-L203C17 has already transferred *all* the available bidding tokens to the beneficiary / auctionieer by `super.settle`. The tokens that you're transferring here dont actually belong to the crowdsale.

If you first were updating the fundingGoal and *then* settle the sale, it'd behave correctly but now you modified the crowdsale without telling anyone. That's technically possible but extremely intransparent since the actual funding goal has been modified as an effect of another function (and nobody gets and update on this since no event is emitted here)

Imo this whole approach is quite debatable, since fees will have an effect on real payouts. Inside the current codebase you cannot solve it elegantly since your parent contract's settle function can't be easily overridden without copying its full code (I added pointcuts for other cases but not for this one). 

With that, I'd rather suggest to find a good pointcut / hook function that computes the amount of tokens that are supposed to be sent and modify the base crowdsale contract's `settle` function accordingly (call that method inside its settle function and return the default fundingGoal, override that in your contract).

We also *could* consider making fees a concept of the base contract itself. You imo can add the fees taken by a crowdsale instance safely *at the end* of the base sale struct, this imo won't break anything else and default to 0 and update the base settle function as described above. That shouldn't break the other derivatives since it only has an effect for the auctioneer, not for the bidders. It's imo also fine to do it since our only existing mainnet contract can't be changed anyway and subgraphs will stay compatible.